ledger/streaming/client/src/block_validator.rs

```diff
@@ -0,0 +1,261 @@
+// Copyright (c) 2018-2022 The MobileCoin Foundation
+
+//! Stream that validates block components
+
+use futures::{future, Stream, StreamExt};
+use mc_common::{
+    logger::{log, Logger},
+    HashSet,
+};
+use mc_ledger_db::Ledger;
+use mc_ledger_streaming_api::{
+    BlockStream, BlockStreamComponents, Error as StreamError, Result as StreamResult,
+};
+use mc_transaction_core::{compute_block_id, ring_signature::KeyImage, BlockID};
+
+/// Create stream factory for validating individual blocks within a stream.
+/// Valid blocks will passed on, blocks that don't pass will pass an error.
+#[derive(Debug, Clone)]
+pub struct BlockValidator<US: BlockStream + 'static, L: Ledger + 'static> {
+    upstream: US,
+    ledger: Option<L>,
+    logger: Logger,
+}
+
+impl<US: BlockStream + 'static, L: Ledger + 'static> BlockValidator<US, L> {
+    /// Create new block validation stream
+    pub fn new(upstream: US, ledger: Option<L>, logger: Logger) -> Self {
+        Self {
+            upstream,
+            ledger,
+            logger,
+        }
+    }
+}
+
+impl<US: BlockStream + 'static, L: Ledger + Clone + 'static> BlockStream for BlockValidator<US, L> {
+    type Stream = impl Stream<Item = StreamResult<BlockStreamComponents>>;
+
+    /// Get block stream that performs validation
+    fn get_block_stream(&self, starting_height: u64) -> StreamResult<Self::Stream> {
+        //get block id from ledger if it exists, else initialize to an empty value
+        let ledger = self.ledger.clone();
+        let prev_block_id = if self.ledger.is_some() && starting_height > 0 {
+            ledger.as_ref().unwrap().get_block(starting_height - 1)?.id
+        } else {
+            BlockID::default()
+        };
+        let additional_key_images: HashSet<KeyImage> = HashSet::default();
+
+        log::info!(self.logger, "Creating block validation stream");
+        let stream = self.upstream.get_block_stream(starting_height)?;
+
+        Ok(stream.scan(
+            (
+                ledger,
+                prev_block_id,
+                additional_key_images,
+                starting_height,
+            ),
+            |state, component| {
+                match component {
+                    Ok(component) => {
+                        let (ledger, prev_block_id, additional_key_images, starting_height) = state;
+
+                        let block = component.block_data.block();
+                        let block_contents = component.block_data.contents();
+
+                        if *starting_height == block.index && ledger.is_none() && block.index > 0 {
```

Comment:
Consider moving this to an `else if` below the ledger check below (currently line 80)